### ISSUE TITLE
refactor Value types

### ISSUE THREAD

**Issue by guangyuz** [zkmove/zkmove#109]
Currently simple values have no bound address. To flatten a simple value we have to pass an address, but to flatten a complex value we don't. This kind of inconsistency can be eliminated by binding the address to the value.

```
pub fn flatten(&self, address: ValueAddress<F>)
```

**Comment by guangyuz** [comment]
lerencao's previous comments for reference.

```
pub enum LocatedValue<F: FieldExt> {
    Invalid,
    U8(ValueAddress<F>, U8<F>),

    U64(ValueAddress<F>, U64<F>),
    U128(ValueAddress<F>, U128<F>),
    Bool(ValueAddress<F>, Bool<F>),
    Address(ValueAddress<F>, Address<F>),
    Container(AddressedContainer<F>),
    ContainerRef(AddressedContainerRef<F>),
    IndexRef(AddressedIndexedRef<F>),
}

#[derive(Debug)]
pub enum AddressedContainer<F: FieldExt> {
    Locals(ValueAddress<F>, Rc<RefCell<Vec<Value<F>>>>),
    Struct(ValueAddress<F>, Rc<RefCell<Vec<Value<F>>>>),
}

pub enum AddressedContainerRef<F: FieldExt> {
    // if we don't need location of inner container, we can just replace it with Container
    Local(ValueAddress<F>, AddressedContainer<F>),
    Global(ValueAddress<F>, AddressedContainer<F>),
}
pub struct AddressedIndexedRef<F: FieldExt> {
    location: ValueAddress<F>,
    index: F,
    // if we don't need location of inner container_ref, we can just replace it with ContainerRef
    container_ref: AddressedContainerRef<F>,
}
impl<F: FieldExt> From<(Value<F>, ValueAddress<F>)> for LocatedValue<F> {
    fn from((c, location): (Value<F>, ValueAddress<F>)) -> Self {
        match c {
            Value::Invalid => LocatedValue::Invalid,
            Value::U8(v) => LocatedValue::U8(location, v),
            Value::U64(v) => LocatedValue::U64(location, v),
            Value::U128(v) => LocatedValue::U128(location, v),
            Value::Bool(v) => LocatedValue::Bool(location, v),
            Value::Address(v) => LocatedValue::Address(location, v),
            Value::Container(Container::Locals(/* this field is dropped*/ _, vs)) => {
                AddressedContainer::Locals(location, vs)
            }
            Value::Container(Container::Struct(/*this field is dropped*/ _, vs)) => {
                AddressedContainer::Struct(location, vs)
            }
            Value::ContainerRef()
        }
    }
}

```

**Comment by nanne007** [comment]
closed by zkmove/zkmove-vm#94
